### PR TITLE
Revert in autosubmit bot

### PR DESCRIPTION
**Pull request by ricardoamador**
## Description

Adding support for the auto-submit bot to temporarily handle revert requests in order to allow OWNER and MEMBER developers to submit revert pull requests immediately without having to wait for reviews and CI check to complete. This will allow us to fix the tree status more quickly than before.

This feature currently supports the method of opening a revert request via the Github UI. Phase 2 will allow a more automated process to revert pull requests. Once the develop has opened the revert request they will need to add a  'revert' label to the pull request and make sure a 'Reverts OWNER/REPO#ID' link (which Github creates if the request was opened via the UI) is present in the pull request in order for the bot to process the revert successfully.

This bot will not attempt to validate the changes in the pull request but will make the following checks on the revert request:
1. The author is a repo OWNER or MEMBER.
2. The pull request is Mergeable as determined by Github (There are no merge conflicts). 
3. The original pull request and the current revert request have the same file sets.

If those conditions are met then the request will be merged automatically without a review and without waiting for the normal CI checks to complete.

This PR has been updated to add the metrics collection through the autosubmit bot. We will now collect data on the type of requests that are made through the autosubmit bot.

*List which issues are fixed by this PR. You must list at least one issue.*

Fixes #108374

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: auto_submit/lib/service/github_service.dart [hunk 1]
```diff
@@ -29,6 +29,36 @@ class GithubService {
     return await github.repositories.getCommit(slug, sha);
   }
 
+  Future<List<PullRequestFile>> getPullRequestFiles(RepositorySlug slug, PullRequest pullRequest) async {
+    int? pullRequestId = pullRequest.number;
+    List<PullRequestFile> listPullRequestFiles = [];
```
Comment (keyonghan) [inline auto_submit/lib/service/github_service.dart:34]: Use `final` keyword for variables that are not expected to change in runtime.

File: auto_submit/lib/service/github_service.dart [hunk 2]
```diff
@@ -29,6 +29,36 @@ class GithubService {
...
+
+    if (pullRequestId == null) {
+      return listPullRequestFiles;
+    }
+
+    Stream<PullRequestFile> pullRequestFiles = github.pullRequests.listFiles(slug, pullRequestId);
```
Comment (keyonghan) [inline auto_submit/lib/service/github_service.dart:40]:
```suggestion
   listPullRequestFiles = await github.pullRequests.listFiles(slug, pullRequestId).toList();
```

File: auto_submit/lib/service/github_service.dart [hunk 3]
```diff
@@ -29,6 +29,36 @@ class GithubService {
...
+  Future<Issue> createIssue(
+    RepositorySlug repositorySlug,
+    String title,
+    String body,
+  ) async {
+    IssueRequest issueRequest = IssueRequest(
+      title: title,
+      body: body,
+    );
+    return github.issues.create(repositorySlug, issueRequest);
```
Comment (keyonghan) [inline auto_submit/lib/service/github_service.dart:59]: This should be async process: `await`.

File: auto_submit/lib/service/validation_service.dart [hunk 4]
```diff
@@ -24,7 +25,14 @@ import 'package:github/github.dart' as github;
 /// a commit to ensure it is ready to land, it has been reviewed, and it has been
 /// tested. The expectation is that the list of validation will grow overtime.
 class ValidationService {
+  Revert? revertValidation;
```
Comment (keyonghan) [inline auto_submit/lib/service/validation_service.dart:28]: nit: move below the constructor.

File: auto_submit/lib/service/validation_service.dart [hunk 5]
```diff
@@ -24,7 +25,14 @@ import 'package:github/github.dart' as github;
...
+
   ValidationService(this.config) {
+    /// revert validation will go here since it needs to be done before the
```
Comment (keyonghan) [inline auto_submit/lib/service/validation_service.dart:31]: Doc format.

File: auto_submit/lib/service/validation_service.dart [hunk 6]
```diff
@@ -88,6 +97,18 @@ class ValidationService {
...
+    /// revert label is the only label present.
+    List<String> labels = (messagePullRequest.labels as List<github.IssueLabel>)
+        .map<String>((github.IssueLabel labelMap) => labelMap.name)
+        .toList();
+
+    // Process a pull request that is marked with the revert label.
+    if (labels.contains(Config.kRevertLabel) && !labels.contains(Config.kAutosubmitLabel)) {
+      processRevertRequest(config, result, messagePullRequest, ackId, pubsub);
+      return;
+    }
```
Comment (keyonghan) [inline auto_submit/lib/service/validation_service.dart:110]: 1) Let's wrap this to a sub function to simplify the logics:
```
if (shouldRevert()) {
  processRevertRequest();
  return;
}
```

This logic sounds a better fit in `processMessage`:
```
if (should revert) {
processRevertRequest;
} else {
ProcessPullRquest;
}
```
2) `messagePullRequest.labels` returns the labels when messaged was pushed to pub/sub. We should obtain the latest labels via an API call.

File: auto_submit/lib/service/validation_service.dart [hunk 7]
```diff
@@ -154,4 +175,46 @@ class ValidationService {
     }
     return true;
   }
+
+  /// The logic for processing a revert request. If the revert request is
```
Comment (keyonghan) [inline auto_submit/lib/service/validation_service.dart:179]: Follow doc format.

File: auto_submit/lib/service/validation_service.dart [hunk 8]
```diff
@@ -154,4 +175,46 @@ class ValidationService {
...
+  /// processed successfully a new tracking review issue is opened as a result
+  /// since we want to be able to follow up with a formal review.
+  ///
+  /// TODO (ricardoamador) move this code to its own service.
+  Future<void> processRevertRequest(
+      Config config, QueryResult result, github.PullRequest messagePullRequest, String ackId, PubSub pubsub) async {
```
Comment (keyonghan) [inline auto_submit/lib/service/validation_service.dart:185]: Nit: add a trailing comma.

File: auto_submit/lib/validations/revert.dart [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+import '../service/log.dart';
+
+class Revert extends Validation {
+  Revert({
+    required Config config,
+  }) : super(config: config);
+
+  static const Set<String> allowedReviewers = <String>{ORG_MEMBER, ORG_OWNER};
+
+  /// Validate a revert pull request.
```
Comment (keyonghan) [inline auto_submit/lib/validations/revert.dart:21]: Nit: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-useless-documentation

File: auto_submit/lib/validations/revert.dart [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+  /// the files in proposer are the same or a subset of proposee then we can
+  /// consider this a revert request.
+  Future<bool> comparePullRequests(
+      github.RepositorySlug repositorySlug, github.PullRequest revert, github.PullRequest current) async {
+    final GithubService githubService = await config.createGithubService(repositorySlug);
+    List<PullRequestFile> originalPullRequestFiles = await githubService.getPullRequestFiles(repositorySlug, revert);
+    List<PullRequestFile> currentPullRequestFiles = await githubService.getPullRequestFiles(repositorySlug, current);
+
+    return validateFileSetsAreEqual(originalPullRequestFiles, currentPullRequestFiles);
+  }
```
Comment (keyonghan) [inline auto_submit/lib/validations/revert.dart:128]: Let's move these logics to githubService.

File: auto_submit/lib/service/validation_service.dart [hunk 11]
```diff
@@ -154,4 +175,46 @@ class ValidationService {
...
+        'Revert request by author ${result.repository!.pullRequest!.author}',
+      );
+
+      log.info('$issue was created to track the review for $prNumber');
+    } else {
+      // since we do not temporarily ignore anything with a revert request we
+      // know we will report the error and remove the label.
+      final int prNumber = messagePullRequest.number!;
+      github.RepositorySlug slug = messagePullRequest.base!.repo!.slug();
+      final GithubService gitHubService = await config.createGithubService(slug);
```
Comment (keyonghan) [inline auto_submit/lib/service/validation_service.dart:209]: Dup logics that can be moved out of the if else.

File: auto_submit/lib/service/config.dart [hunk 12]
```diff
@@ -173,6 +174,10 @@ class Config {
   /// The label which shows the overrideTreeStatus.
   String get overrideTreeStatusLabel => 'warning: land on red to fix tree breakage';
 
+  /// The autosubmit label.
+  String get autosubmitLabel => 'autosubmit';
+  String get revertLabel => 'revert';
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply auto_submit/lib/service/config.dart:179]: Removed.

File: auto_submit/lib/requests/github_webhook.dart [hunk 13]
```diff
@@ -53,13 +54,21 @@ class GithubWebhook extends RequestHandler {
     }
 
     final PullRequest pullRequest = PullRequest.fromJson(body['pull_request'] as Map<String, dynamic>);
-    hasAutosubmit = pullRequest.labels!.any((label) => label.name == Config.kAutosubmitLabel);
+    hasAutosubmit = pullRequest.labels!.any((label) => label.name == config.autosubmitLabel);
+    hasRevertLabel = pullRequest.labels!.any((label) => label.name == config.revertLabel);
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply auto_submit/lib/requests/github_webhook.dart:58]: Fixed.

File: auto_submit/lib/requests/github_webhook.dart [hunk 14]
```diff
@@ -53,13 +54,21 @@ class GithubWebhook extends RequestHandler {
...
 
     if (hasAutosubmit) {
       log.info('Found pull request with auto submit label.');
       await pubsub.publish('auto-submit-queue', pullRequest);
     }
 
+    // in order to revert we must not have the autosubmit label. The revert label will
+    // yield to the autosubmit label.
+    if (!hasAutosubmit && hasRevertLabel) {
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply auto_submit/lib/requests/github_webhook.dart:67]: Okay no problem. We can do that.

File: auto_submit/lib/validations/revert.dart [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+    github.PullRequest requestToRevert = await getPullRequest(repositorySlug, pullRequestId);
+
+    // Compare the changes made with the linked pull request.
+    bool requestsMatch = await comparePullRequests(repositorySlug, requestToRevert, messagePullRequest);
+
+    // if the changes are a revert then approve the pull request.
+    if (requestsMatch) {
+      // create a follow on issue to track the review request for this revert.
+      return ValidationResult(
+          true, Action.REMOVE_LABEL, 'Revert request has been verified and will be queued for merge.');
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply auto_submit/lib/validations/revert.dart:66]: Fixed.

File: auto_submit/lib/validations/revert.dart [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    if (matches.isNotEmpty) {
+      // return the first group
+      return matches.elementAt(0).group(1);
+    }
+    return null;
+  }
+
+  /// Split a reverts link on the '#' then the '/' to get the parts of the repo
+  /// slug.
+  /// It is assumed that the link has the format flutter/repo#id.
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply auto_submit/lib/validations/revert.dart:97]: Fixed.

File: auto_submit/original_commit.github.json [hunk 17]
```diff
@@ -0,0 +1,302 @@
+[
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply auto_submit/original_commit.github.json:1]: No they are not. I don't know how they ended up there. I remember specifically deleting them. Removing them.

File: auto_submit/lib/service/github_service.dart (hunk 1 above)
Comment (ricardoamador) [reply auto_submit/lib/service/github_service.dart:34]: Done.

File: auto_submit/lib/service/github_service.dart (hunk 3 above)
Comment (ricardoamador) [reply auto_submit/lib/service/github_service.dart:59]: Fixed.

File: auto_submit/lib/service/validation_service.dart (hunk 4 above)
Comment (ricardoamador) [reply auto_submit/lib/service/validation_service.dart:28]: Done.

File: auto_submit/lib/service/validation_service.dart (hunk 7 above)
Comment (ricardoamador) [reply auto_submit/lib/service/validation_service.dart:179]: Done.

File: auto_submit/lib/service/validation_service.dart (hunk 8 above)
Comment (ricardoamador) [reply auto_submit/lib/service/validation_service.dart:185]: Done.

File: auto_submit/lib/validations/revert.dart (hunk 9 above)
Comment (ricardoamador) [reply auto_submit/lib/validations/revert.dart:21]: Cleaned up all docs in this file.

File: auto_submit/lib/service/validation_service.dart (hunk 11 above)
Comment (ricardoamador) [reply auto_submit/lib/service/validation_service.dart:209]: Fixed.

File: auto_submit/lib/service/validation_service.dart (hunk 6 above)
Comment (ricardoamador) [reply auto_submit/lib/service/validation_service.dart]: Done.

File: auto_submit/test/validations/revert_test_data.dart [hunk 18]
```diff
@@ -0,0 +1,1703 @@
+// Copyright 2022 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply auto_submit/test/validations/revert_test_data.dart:2]: No problem.

File: auto_submit/lib/validations/revert.dart [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+    bool requestsMatch = await comparePullRequests(repositorySlug, requestToRevert, messagePullRequest);
+
+    if (requestsMatch) {
+      return ValidationResult(true, Action.IGNORE_FAILURE, 'Revert request has been verified and will be queued for merge.');
+    }
+
+    return ValidationResult(false, Action.REMOVE_LABEL, 'Validation of the revert request has failed. Verify the files in the revert request are the same as the original PR and resubmit the revert request.');
+  }
+
+  /// Only a team member and code owner can submit a revert request with a review.
```
Comment (zanderso) [inline auto_submit/lib/validations/revert.dart:61]:
```suggestion
  /// Only a team member and code owner can submit a revert request without a review.
```

**Review by zanderso — COMMENTED** [review]
Thanks for doing this.

File: auto_submit/test/validations/revert_test_data.dart (hunk 18 above)
[the remark this answers is not in the document]
Comment (ricardoamador) [reply auto_submit/test/validations/revert_test_data.dart:2]: Done.

File: auto_submit/lib/validations/revert.dart (hunk 10 above)
Comment (ricardoamador) [reply auto_submit/lib/validations/revert.dart]: Done.

File: auto_submit/lib/validations/revert.dart (hunk 19 above)
Comment (ricardoamador) [reply auto_submit/lib/validations/revert.dart:62]: Done!

**Comment by ricardoamador** [comment]
@zanderso Of course! I would like to get your input further down the road when I start the standalone service.

File: auto_submit/lib/validations/revert.dart [hunk 20]
```diff
@@ -0,0 +1,98 @@
+// Copyright 2022 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:auto_submit/model/auto_submit_query_result.dart' as auto;
+import 'package:auto_submit/validations/validation.dart';
+import 'package:github/github.dart' as github;
+import 'package:auto_submit/service/github_service.dart';
```
Comment (keyonghan) [inline auto_submit/lib/validations/revert.dart:8]: Nit: sort the imports alphabetically. Please check other imports as well.

File: auto_submit/lib/validations/revert.dart [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+    if (revertLink == null) {
+      String message =
+          'A reverts link could not be found or was formatted incorrectly. Format is \'Reverts owner/repo#id\'';
+      log.info(message);
+      return ValidationResult(false, Action.REMOVE_LABEL, message);
+    }
+
+    github.RepositorySlug repositorySlug = _getSlugFromLink(revertLink);
+    GithubService githubService = await config.createGithubService(repositorySlug);
+    int pullRequestId = _getPullRequestIdFromLink(revertLink);
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply auto_submit/lib/validations/revert.dart:51]: I don't think this is the same here as this is the original closed pull requests id number. Not the current open revert request id.

File: auto_submit/lib/service/github_service.dart (hunk 2 above)
Comment (keyonghan) [reply auto_submit/lib/service/github_service.dart:40]: Seems the changes of this is not in?

**Review by keyonghan — APPROVED** [review]
LGTM with a nit. 
Thanks for the contribution!
Let's take a close monitoring the GCP logs the following morning after this lands.

**Comment by keyonghan** [comment]
Let's separate the bigquery logic to a separate PR for easier review and tracking.

**Comment by ricardoamador** [comment]
Sure, no problem.

**Pull request merged**